research/deep_speech/deep_speech_model.py

```diff
@@ -0,0 +1,174 @@
+  rnn_layer = rnn_cell(
+      rnn_hidden_size, activation=rnn_activation, return_sequences=True,
+      name="rnn_{}".format(layer_id))
+  if is_bidirectional:
+    rnn_layer = tf.keras.layers.Bidirectional(rnn_layer, merge_mode="sum")
+
+  rnn_output = rnn_layer(input_data)
+  return rnn_output
+
+
+def _ctc_lambda_func(args):
+  """Compute ctc loss."""
+  y_pred, labels, input_length, label_length = args
+  return tf.keras.backend.ctc_batch_cost(
+      labels, y_pred, input_length, label_length)
+
+def _cal_ctc_input_length(args):
+  """Compute the actual input length after convolution.
+
+  Basically, we need to know the scaled input_length after conv layers.
+  new_input_length = old_input_length * ctc_time_steps / max_time_steps
+  """
+  input_length, input_data, y_pred= args
+  max_time_steps = tf.shape(input_data)[1]
+  ctc_time_steps = tf.shape(y_pred)[1]
+  ctc_input_length = tf.multiply(
+    tf.to_float(input_length), tf.to_float(ctc_time_steps))
+  ctc_input_length = tf.to_int32(tf.floordiv(
+    ctc_input_length, tf.to_float(max_time_steps)))
+  return ctc_input_length
+
+
+class DeepSpeech(tf.keras.models.Model):
+  """DeepSpeech model."""
+
+  def __init__(self, input_shape, num_rnn_layers, rnn_type, is_bidirectional,
+               rnn_hidden_size, rnn_activation, num_classes, use_bias):
+    """Initialize DeepSpeech2 model.
+    Args:
+      input_shape: an tuple to indicate the dimension of input dataset. It has
+        the format of [time_steps(T), feature_bins(F), channel(1)]
+      num_rnn_layers: an integer, the number of rnn layers. By default, it's 5.
+      rnn_type: a string, one of the supported rnn cells: gru, rnn and lstm.
+      is_bidirectional: a boolean to indicate if the rnn layer is bidirectional.
+      rnn_hidden_size: an integer for the number of hidden states in each unit.
+      rnn_activation: a string to indicate rnn activation function. It can be
+        one of tanh and relu.
+      num_classes: an integer, the number of output classes/labels.
+      use_bias: a boolean specifying whether to use bias in the last fc layer.
+    """
+    # Input variables
+    input_data = tf.keras.layers.Input(
+        shape=input_shape, name="features")
+
+    # Two cnn layers
+    conv_layer_1 = _conv_bn_layer(
+        input_data, filters=32, kernel_size=(41, 11), strides=(2, 2),
+        layer_id=1)
+    print("conv_layer_1", conv_layer_1)
+
+    conv_layer_2 = _conv_bn_layer(
+        conv_layer_1, filters=32, kernel_size=(21, 11), strides=(2, 1),
+        layer_id=2)
+    # output of conv_layer2 with the shape of
+    # [batch_size (N), times (T), features (F), channels (C)]
+
+    # RNN layers.
+    # Convert the conv output to rnn input
+    rnn_input = tf.keras.layers.TimeDistributed(tf.keras.layers.Flatten())(
+        conv_layer_2)
+
+    rnn_cell = _SUPPORTED_RNNS[rnn_type]
+    for layer in xrange(num_rnn_layers):
+      is_batch_norm = (layer != 0)  # No batch normalization on the first layer
```

Comment:
nit: layer sounds like the Layer obj. Can we call this layer_counter or i or something that makes it more clearly an incrementer?